src/ITensorNetworks/peps.jl

```diff
@@ -0,0 +1,142 @@
+  tensor_grid = Matrix{ITensor}(undef, Ny, Nx)
+  # we assume the PEPS at least has size (2,2). Can generalize if necessary
+  @assert(Nx >= 2 && Ny >= 2)
+
+  lh = Matrix{Index}(undef, Ny, Nx - 1)
+  for ii in 1:(Nx - 1)
+    for jj in 1:(Ny)
+      lh[jj, ii] = Index(linkdims, "Lh,$jj,$ii")
+    end
+  end
+  lv = Matrix{Index}(undef, Ny - 1, Nx)
+  for ii in 1:(Nx)
+    for jj in 1:(Ny - 1)
+      lv[jj, ii] = Index(linkdims, "Lv,$jj,$ii")
+    end
+  end
+
+  # boundary cases
+  tensor_grid[1, 1] = emptyITensor(T, lh[1, 1], lv[1, 1], sites[1, 1])
+  tensor_grid[1, Nx] = emptyITensor(T, lh[1, Nx - 1], lv[1, Nx], sites[1, Nx])
+  tensor_grid[Ny, 1] = emptyITensor(T, lh[Ny, 1], lv[Ny - 1, 1], sites[Ny, 1])
+  tensor_grid[Ny, Nx] = emptyITensor(T, lh[Ny, Nx - 1], lv[Ny - 1, Nx], sites[Ny, Nx])
+  for ii in 2:(Nx - 1)
+    tensor_grid[1, ii] = emptyITensor(T, lh[1, ii], lh[1, ii - 1], lv[1, ii], sites[1, ii])
+    tensor_grid[Ny, ii] = emptyITensor(
+      T, lh[Ny, ii], lh[Ny, ii - 1], lv[Ny - 1, ii], sites[Ny, ii]
+    )
+  end
+
+  # inner sites
+  for jj in 2:(Ny - 1)
+    tensor_grid[jj, 1] = emptyITensor(T, lh[jj, 1], lv[jj, 1], lv[jj - 1, 1], sites[jj, 1])
+    tensor_grid[jj, Nx] = emptyITensor(
+      T, lh[jj, Nx - 1], lv[jj, Nx], lv[jj - 1, Nx], sites[jj, Nx]
+    )
+    for ii in 2:(Nx - 1)
+      tensor_grid[jj, ii] = emptyITensor(
+        T, lh[jj, ii], lh[jj, ii - 1], lv[jj, ii], lv[jj - 1, ii], sites[jj, ii]
+      )
+    end
+  end
+
+  return PEPS(tensor_grid)
+end
+
+PEPS(sites::Matrix{<:Index}, args...; kwargs...) = PEPS(Float64, sites, args...; kwargs...)
+
+PEPS(data::Array, size::Tuple{Int64,Int64}) = PEPS(reshape(data, size[1], size[2]))
+
+function randomizePEPS!(P::PEPS)
+  dimy, dimx = size(P.data)
+  for ii in 1:dimx
+    for jj in 1:dimy
+      randn!(P.data[jj, ii])
+      normalize!(P.data[jj, ii])
+    end
+  end
+end
+
+function Base.:+(A::PEPS, B::PEPS)
+  @assert(size(A.data) == size(B.data))
+  out = [t1 + t2 for (t1, t2) in zip(vcat(A.data...), vcat(B.data...))]
+  return PEPS(out, size(A.data))
+end
+
+function Base.:-(A::PEPS, B::PEPS)
+  @assert(size(A.data) == size(B.data))
+  out = [t1 - t2 for (t1, t2) in zip(vcat(A.data...), vcat(B.data...))]
+  return PEPS(out, size(A.data))
+end
+
+function Base.:*(c::Number, A::PEPS)
+  out = [c * t for t in vcat(A.data...)]
+  return PEPS(out, size(A.data))
+end
+
+function Base.:*(A::PEPS, B::PEPS)
+  @assert(size(A.data) == size(B.data))
+  out_scalar = 0
+  for (t1, t2) in zip(vcat(A.data...), vcat(B.data...))
+    out_scalar += ITensors.scalar(t1 * t2)
+  end
+  return out_scalar
+end
+
+get_prime_bonds(indices; ham=true) = ham ? indices : indices[1:(end - 1)]
+
+function ITensors.prime(peps::PEPS; ham=true)
+  prime_peps = map(
+    tensor -> prime(tensor, get_prime_bonds(inds(tensor); ham=ham)), peps.data
+  )
+  return PEPS(prime_peps)
+end
+
+# Get the tensor network of <peps|peps'>
+function inner_network(peps::PEPS, peps_prime::PEPS)
```

Comment:
Couldn't this function do the priming internally?

Reply:
I think I answered my own question by looking at the gradient definitions below. I guess it is organized this way so you don't have to define pullbacks for `inner_network` and `generate_inner_network`?

I guess it is ok for now as long as it makes the gradients work. As an external interface this would seem a bit off to me, but I suppose it is more for internal usage to set things up to perform the gradient.

Reply:
Note that we don't want to differentiate any tensor network graphs, since those are only input to AutoHOOT to calculate correct gradients. We can definitely do something like 
```julia
function inner_network(peps::PEPS)
    peps_prime = prime(peps)
    net = network(peps_prime, peps)
end
```
but we still have to make sure the internal function network is nondifferentiable, which I think is just the same as what we are doing now

Reply:
If another goal is to just get the AD chainrules for ITensors working I agree that this seems not a very good choice. But for AutoHOOT I didn't have other better ideas on the design.

Reply:
Yes, I just thought that something like the 

> Note that we don't want to differentiate any tensor network graphs, since those are only input to AutoHOOT to calculate correct gradients. We can definitely do something like
> 
> ```julia
> function inner_network(peps::PEPS)
>     peps_prime = prime(peps)
>     net = network(peps_prime, peps)
> end
> ```
> 
> but we still have to make sure the internal function network is nondifferentiable, which I think is just the same as what we are doing now

Yes, I was just pointing out that I think something like this new definition of `inner_network` makes more sense to me, purely from a design perspective. Though I might call this something like `norm2_network` or `sqrnorm_network`.

Reply:
Actually, I think another reason (and I felt it's also the major reason) for my making the peps_prime as an input to this function is that, this function is currently called inside of generate_inner_network, which is non-differentiable. But we do want the prime function to be differentiable. which is why I call it outside: https://github.com/LinjianMa/ITensorNetworkAD.jl/blob/main/src/Optimizations/peps.jl#L87.

A better way to understand the flow is that we are constructing the loss function using the following flow.
1. define all variables, this includes all potential prime funcitons
2. define tensor network graphs, this function should be nondifferentiable
3. call AutoHOOT to get differentiations.

Reply:
I see. I guess I was picturing just two steps:

1. Set up all unique tensors in the network(s) (with priming/tagging to make indices unique).
2. Call AutoHOOT on the total network(s) to get the gradient.

And I was picturing both steps could be made fully differentiable, Step 1 with the ITensor ChainRules definitions and general Zygote code, and Step 2 through network level ChainRules that make use of AutoHOOT. Why is there a non-differentiable step? By "define tensor network graphs", is that just collecting all of the unique tensors of the networks together, or is there something more to it then that?

Reply:
That is a list of networks, so something like `[[A,B,C], [A,B,D], [B,C,D]]`.
The step to construct this expression should not be differentiable.

Reply:
I thought those kinds of operations (like collecting various networks into a big network) would be differentiable, like:
```julia
julia> using Zygote

julia> f(x) = (y = [[x^2 sin(x)] [x^3 cos(x)]]; sum(y[1]) + sum(y[2]))
f (generic function with 1 method)

julia> f(4)
15.243197504692072

julia> f'(4)
7.346356379136388
```
This is of course a much simpler example, but fundamentally it seems like it just involves collecting tensors together into vectors in some way (but obviously more complicated).

Is the issue that you see an error when you try to differentiate through collecting the tensor network graph (in `generate_inner_network`), or there is some more fundamental reason why you don't think that step should be differentiated? I'm just trying to understand better, since you understand these things better than I do. I would have thought that to track the derivatives you would have to differentiate through `generate_inner_network` so it is just surprising to me that you seem to not have to at all. We can also chat about it in our next meeting since it might be too involved to explain here.

Reply:
So there are major two reasons:
1. the differentiation of generate_inner_network is not what we want. The list of tensor network graphs is the first input to the function batch_tensor_contraction https://github.com/ITensor/ITensorNetworkAD.jl/blob/main/src/ITensorAutoHOOT/batch_contraction.jl#L83, and we only want differentiations to the variables.
2. since the differentiation of generate_inner_network is useless, we'd better just set the function to be nondifferentiable. Otherwise it could be the case that within the function it calls some other non-differentiable functions, such as using hashmaps, which will yield error.

Reply:
I think that makes sense, I'll have to think about it a bit. Thanks for explaining.